_posts/2020-09-19-github-actions-with-faasd.md

```diff
@@ -0,0 +1,256 @@
+
+* faas-cli
+
+faas-cli is a command line tool used to interact with OpenFaaS, from deploying your functions to viewing logs.
+Follow the installation instructions [here](https://docs.openfaas.com/cli/install/)
+
+> You can follow along with the code used in this tutorial from this [repo](https://github.com/utsavanand2/faasd-hello)
+
+## Create a Droplet with faasd
+
+Create a DigitalOcean Droplet with a Ubuntu distribution, with any plan according to what you want to run, faasd being light weight I'm going to use the $5 basic plan.
+
+![Choose a distribution and a plan](/images/2020-09-19-github-actions-faasd/choose-distro.jpg)
+
+Under `Select additional options`, select `User data` and paste in the contents of the cloud-config file from [here](https://github.com/openfaas/faasd/blob/master/cloud-config.txt)
+
+![Select additional options](/images/2020-09-19-github-actions-faasd/additional-options-cloud-config.jpg)
+
+
+![faasd droplet](/images/2020-09-19-github-actions-faasd/faasd-droplet.jpg)
+
+Within as little as 30 seconds, your faasd instance will be up and running.
+
+## Get the OpenFaaS gateway password
+
+Get the gateway password by running the following commands
+
+```sh
+export IP="143.110.184.36"
+export PASSWORD=$(ssh root@$IP "sudo cat /var/lib/faasd/secrets/basic-auth-password")
+```
+
+Assuming that you have faas-cli already installed, login to faasd with faas-cli with the following commands
+
+```sh
+export OPENFAAS_URL=http://$IP:8080
+echo $PASSWORD | faas-cli login --password-stdin -u admin -g $OPENFAAS_URL
+```
+
+## Create an OpenFaaS function
+
+Now that we've logged into faasd with faas-cli, we're ready to run our functions.
+I'm going to deploy a golang function for this tutorial.
+
+```sh
+# create a new directory for our project
+mkdir faasd-hello
+# pull golang-http templates
+faas-cli template store pull golang-http
+# create a new function with faas-cli
+faas-cli new faasd-hello --lang golang-http --prefix utsavanand2
+```
+
+If you look into the root directory of the project, faas-cli has created two additional files:
+
+```
+./faasd-hello/handler.go
+./faasd-hello.yml
+```
+
+The content of our handler looks something like this:
+![./faasd-hello/handler.go](/images/2020-09-19-github-actions-faasd/handler-go.jpg)
+
+The YAML file is used to configure the deployment and the build and runtime environment of the function
+~[./faasd-hello.yml](/images/2020-09-19-github-actions-faasd/faasd-hello-yaml.jpg)
+
+## Setup a workflow file for Github Actions
+
+In the root directory of the project run:
+
+```sh
+mkdir -p .github/workflows
+nano ./github/workflows/main.yml
+```
+
+And paste in the following contents into main.yml file
+
+```yaml
+name: CI-CD
+
+on:
+  push:
+    branches: 
+      - master
+
+jobs:
+  func-build:
+    runs-on: ubuntu-latest
+    steps:
+      - 
+        name: Checkout
+        uses: actions/checkout@v2
+      -
+        name: Pull template
```

Comment:
can this be simplified to this

```yaml
- name: Pull template
  uses: docker://openfaas/faas-cli:latest
  run: faas-cli template store pull golang-http
```

Per these docs https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-a-docker-hub-action

Since it has an entrypoint in that image it might be

```yaml
- name: Pull template
  uses: docker://openfaas/faas-cli:latest
  run: template store pull golang-http
```

Reply:
Based on this https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepswithargs 

it seems like you can also do something like

```yaml
- name: Pull template
  uses: docker://openfaas/faas-cli:latest
  with:
    entrypoint: faas-cli
    args: template store pull golang-http
```

Reply:
Thanks @LucasRoesler! This looks way more simplified. I'll try it out now.

Reply:
It looks like we need to mount the volume inside our docker container to the runner's directory, for it to detect the pulled templates, I fear we'll face the same problem when we use `faas login`. This was the reason why I had to mount the volumes for the build to be sucessful, and if we go down that path, then all of those `chown` commands will have to be used again.
<img width="1668" alt="Screenshot 2020-09-23 at 10 38 02 PM" src="https://user-images.githubusercontent.com/25264581/94046288-1c79a780-fdee-11ea-89ab-d8490ae17b32.png">
I also found [this](https://github.community/t/bind-mount-additional-volumes-for-image-in-github-action/17867) comment telling it's not possible at the moment.
Are you aware of any options that might help us with this @LucasRoesler?
If not do you think we should just use the default CLI and use the `docker` with `chown` only where we need to build the function image?

Reply:
We can also fork your current [OpenFaaS Github Action](https://github.com/LucasRoesler/openfaas-action) to the official OpenFaaS repo and try to keep it maintained with new faas-cli releases?

Reply:
Forking and updating an official action would be pretty nice, especially if we can design it so that we just pass arguments to faas-cli instead of trying to create a action for every command

This approach of installing the cli could also be really good https://github.com/marketplace/actions/setup-hub